File: open/NVIDIA/code/gnmt/tensorrt/GNMT.py
# Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import os, sys
import shutil
sys.path.insert(0, os.getcwd())
from code.common import logging, dict_get, run_command

g_calibration_cache = "code/gnmt/tensorrt/data/Int8CalibrationCache"

class GNMTBuilder():

    def __init__(self, args):
        self.args = args
        self.name = "gnmt"
        self.system_id = args["system_id"]
        self.scenario = args["scenario"]
        self.engine_dir = "./build/engines/{:}/{:}/{:}".format(self.system_id, self.name, self.scenario)
        self.device_type = "gpu"

        self.precision = args["precision"]
        self.precision_flag = ""
        if (self.precision == "int8") or (self.precision == "fp16"):
            self.precision_flag = "-t fp16"
        if dict_get(args, "enable_int8_generator", default=False):
            self.precision_flag += " --int8Generator --calibration_cache {}".format(g_calibration_cache)

    def build_engines(self):
        if self.scenario == "MultiStream":
            raise NotImplementedError("GNMT MultiStream scenario is not yet implemented")
        elif self.scenario == "SingleStream":
            batch_sizes = [ self.args["batch_size"] ]
        elif self.scenario == "Offline":
            batch_sizes = [ self.args["batch_size"] ]
        elif self.scenario == "Server":
            batch_sizes = self.args["batch_sizes"]
        else:
            raise ValueError("Invalid scenario: {:}".format(self.scenario))

        beam_size = dict_get(self.args, "beam_size", default=10)
        seq_len_slots = dict_get(self.args, "seq_len_slots", default=1)
        mpbs_flag = ""
        if "max_persistent_bs" in self.args:
            mpbs_flag = "--max_persistent_bs {:}".format(self.args["max_persistent_bs"])

        if not os.path.exists(self.engine_dir):
            os.makedirs(self.engine_dir)

        for batch_size in batch_sizes:
            engine_name = "{:}/{:}-{:}-gpu-b{:}-{:}.plan".format(self.engine_dir, self.name, self.scenario, batch_size, self.precision)
            logging.info("Building {:}".format(engine_name))


            cmd = "build/bin/GNMT/gnmt --seq_len_slots {seq_len_slots} --bm {beam_size} --build_only --bs {batch_size} --store_engine {engine_name} {max_persistent_bs} {precision}".format(
                seq_len_slots=seq_len_slots,
                beam_size=beam_size,
                batch_size=batch_size,
                engine_name=engine_name,
                max_persistent_bs=mpbs_flag,
                precision=self.precision_flag,
            )
            run_command(cmd)

    def calibrate(self):
        beam_size = 1
        batch_size = 1
        num_batches = 64
        output_dir_1 = "calib_phase_1"
        output_dir_2 = "calib_phase_2"

        try:
            phase_1 = "./build/bin/GNMT/gnmt --calibration_phase 1 --input_file {input} --num_batches {num_batches} --bm {beam_size} --bs {batch_size} --output_dir {output_dir}".format(
                    input="build/inference/calibration/translation/calibration_data.tok.bpe.en",
                    num_batches = num_batches,
                    beam_size = beam_size,
                    batch_size = batch_size,
                    output_dir = output_dir_1
                    )

            run_command(phase_1)

            # Get the directory to raw output
            calib_data = None
            for d in os.listdir(output_dir_1):
                if "gnmt_tensors" in d:
                    calib_data = os.path.join(output_dir_1, d)
            assert(calib_data != None)

            phase_2 = "./build/bin/GNMT/gnmt --calibration_phase 2 --bm {beam_size} --calibration_data {input}  --calibration_cache {calib_cache} --output_dir {output_dir}".format(
                    input = calib_data,
                    beam_size = beam_size,
                    batch_size = batch_size,
                    calib_cache = g_calibration_cache,
                    output_dir = output_dir_2
                    )

            run_command(phase_2)

        # Remove the generated intermediate output files
        finally:
            if os.path.exists(output_dir_1):
                shutil.rmtree(output_dir_1)
            if os.path.exists(output_dir_2):
                shutil.rmtree(output_dir_2)

File: open/NVIDIA/code/gnmt/tensorrt/src/GNMT.cpp
/*
 * Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <fstream>
#include <iostream>

#include "GNMTCore.h"
#include "GNMTIndexer.h"
#include "utils.h"
#include "params.h"
#include "plugin/dumpTensorPlugin.h"


int main(int argc, const char** argv)
{

    Params params;

    params.parseOptions(argc, argv);
    if (!params.validateArgs()){
        std::cerr << "Parameter check failed. Exiting" << std::endl;
        exit(-1);
    }

    std::shared_ptr<Config> config(nullptr);

    if(params.shouldLoadConfigFromFile()){
        config.reset(new Config(params.getConfigFile()));
    }
    else{
        config.reset(new Config(params));
    }

    DumpTensorPlugin::setDirName(params.debugDirName.c_str());

    // Read input text
    std::string line;
    ifstream inputEnFile(params.inputFile);
    if (!inputEnFile)
    {
        cout << "Error opening input file " << params.inputFile << endl;
        exit(EXIT_FAILURE);
    }
    std::vector<std::string> enSentence;    //! Vector that contains entire corpus text
    while (std::getline(inputEnFile, line))
    {
        enSentence.push_back(line);
    }

    // Calculate batch related parameters
    int maxBatches = ceil((float) enSentence.size() / params.batchSize);

    // Calculate number of batches and last batch number
    if (params.nbBatch < 0){
        params.nbBatch = maxBatches;
    }
    params.nbBatch = min(maxBatches - params.startBatch, params.nbBatch);
    params.stopBatch = params.startBatch + params.nbBatch;

    // Config for engine
    config->printConfig();

    // Build GNMTCore
    GNMTCore gnmt(config, params.loadEngine, params.storeEngine, params.engineDir, params.profile);

    // In calibrationPhase 2, we only want to build the calibration cache, and don't want to run the main code path
    if (params.calibrationPhase == 2){
        gnmt.buildCalibrationCache();
        return 0;
    }
    
    // Processing parameters
    cout << "Actual Batch Size: " << params.batchSize << ", num  batches: " << params.nbBatch
        << " (start @ " << params.startBatch << " end @ " << params.stopBatch - 1 << ")"
        << ", Input file: " << params.inputFile << ", Output file: " << params.outFile << endl;

    // Setup GNMT
    gnmt.setup();

    if (params.build_only)
        return 0;


    // Process input document batch by batch
    ofstream of(params.outFile);
    stringstream translatedText;

    double t0 = seconds();

    std::vector <std::string> currentSamples (params.batchSize);

    for (RuntimeInfo::currentBatch = params.startBatch; RuntimeInfo::currentBatch < params.stopBatch; RuntimeInfo::currentBatch++)
    {
        // Prepare the batch
        auto startSample = RuntimeInfo::currentBatch * params.batchSize;
        auto endSample = min( (RuntimeInfo::currentBatch+1) * params.batchSize, (int) enSentence.size());
        int actualBatchSize = endSample-startSample;

        // TBD: USE POINTER ARITHMETIC HERE to avoid copying (use vector<string*> enSentence and vector<string*> currentSamples)
        std::copy(&enSentence[startSample], &enSentence[endSample], currentSamples.begin());
        currentSamples.resize(actualBatchSize);   //! Shrink if needed.

        std::vector<std::vector<std::string>> tokenWords = gnmt.translate(currentSamples, !params.disableBatchCulling);

        // Write predicted German sentences to output file
        writeTokenizedBatch(translatedText, tokenWords);
    }

    of << translatedText.str();

    double translation_time = seconds() - t0;
    double translation_speed = enSentence.size()/translation_time;
    printf("\nGNMT takes %.3fs \n", translation_time);
    printf("\nTranslation speed is %.2f sentences/sec\n", translation_speed);

    gnmt.reportAggregateEngineTime();

    return 0;
}
